### ISSUE TITLE
[MI200][Reduction] Fails when all dimensions of a destination tensor are 1

### ISSUE THREAD

**Issue by densamoilov** [ROCmSoftwarePlatform/MIOpen#1705]
When dimensions of source tensor are {15, 12, 3, 5} and dimensions of destination tensor are {1, 1, 1, 1} I get the following runtime error:
```bash
gridwise_generic_reduction_second_call_threadwise_reduce_all_dims.cpp:45:32: error: use of undeclared identifier 'CK_PARAM_TOREDUCE_DIMS'
using toReduceDims  = Sequence<CK_PARAM_TOREDUCE_DIMS>;
                               ^
gridwise_generic_reduction_second_call_threadwise_reduce_all_dims.cpp:46:32: error: use of undeclared identifier 'CK_PARAM_INVARIANT_DIMS'
using invariantDims = Sequence<CK_PARAM_INVARIANT_DIMS>; // this could be empty
                               ^
2 errors generated when compiling for gfx90a.
MIOpen Error: /long_pathname_so_that_rpms_can_package_the_debug_info/data/driver/MLOpen/src/tmp_dir.cpp:45: Can't execute cd /tmp/miopen-gridwise_generic_reduction_second_call_threadwise_reduce_all_dims.cpp-b189-3965-5db3-f0ee;  /opt/rocm-5.0.0/llvm/bin/clang++  --std=c++17 -DCK_AMD_GPU_GFX90A -DCK_USE_AMD_BUFFER_ATOMIC_FADD=0 -DCK_BLOCK_SYNC_LDS_WITHOUT_SYNC_VMEM=1 -DCK_USE_AMD_BUFFER_ADDRESSING=1 -DCK_PARAM_SRC_DATATYPE=1 -DCK_PARAM_DST_DATATYPE=1 -DCK_PARAM_REDUCE_COMPTYPE=1  -DCK_PARAM_BLOCKSIZE=256 -DCK_PARAM_THREAD_BUFFER_LENGTH=8 -DCK_PARAM_ACCESSES_PER_THREAD_INBLOCK=2 -DCK_PARAM_ACCESSES_PER_THREAD_INWARP=2 -DCK_PARAM_REDUCE_OP=3 -DCK_PARAM_NAN_PROPAGATE=0 -DCK_PARAM_REDUCE_INDICES=0 -DCK_PARAM_IN_DIMS=4 -DCK_PARAM_OUT_DIMS=1 -DCK_PARAM_SRC2D_PADDING=1 -DCK_PARAM_DST1D_PADDING=1 -mcpu=gfx90a -Wno-everything --cuda-gpu-arch=gfx90a:xnack- --cuda-device-only -c -O3  -Wno-unused-command-line-argument -I. -mllvm -amdgpu-early-inline-all=true -mllvm -amdgpu-function-calls=false -x hip -isystem /opt/rocm-5.0.0/hip/../include -isystem /opt/rocm-5.0.0/llvm/lib/clang/14.0.0/include/.. -D__HIP_PLATFORM_HCC__=1 -D__HIP_PLATFORM_AMD__=1 -isystem /opt/rocm-5.0.0/hip/include -isystem /opt/rocm-5.0.0/include --hip-link  -DHIP_PACKAGE_VERSION_FLAT=5000022042 gridwise_generic_reduction_second_call_threadwise_reduce_all_dims.cpp -o /tmp/miopen-gridwise_generic_reduction_second_call_threadwise_reduce_all_dims.cpp-b189-3965-5db3-f0ee/gridwise_generic_reduction_second_call_threadwise_reduce_all_dims.cpp.o
```
I noticed that if I change only 1 dimension in the destination tensor to be equal that in the source tensor the issue goes away. Also, other cases with different shapes of source tensor can pass, e.g.  {5, 3, 2, 9}.

Is there a workaround for this issue? The current workaround seems to be avoiding the cases when all dimensions of the destination tensor are 1.

MIOpen 2.16
ROCm 5.0.0

<details>
  <summary>Reproducer (clickable)</summary>

```cpp
#include <functional>
#include <numeric>

#include "miopen/miopen.h"

#define CHECK_MIOPEN(f, msg) \
    do { \
        const auto ___status___ = f; \
        if (___status___ != miopenStatusSuccess) { \
            throw std::runtime_error(msg); \
        } \
    } while (0)
#define CHECK_HIP(f, msg) \
    do { \
        const auto ___status___ = f; \
        if (___status___ != hipSuccess) { throw std::runtime_error(msg); } \
    } while (0)

int main() {
    auto status_miopen = miopenStatusSuccess;
    auto status_hip = hipSuccess;

    const int ndims = 4;
    //int src_dims[ndims] = {5, 3, 2, 9};
    int src_dims[ndims] = {15, 12, 3, 5};
    int dst_dims[ndims] = {1, 1, 1, 1};
    //int src_strides[ndims] = {54, 18, 9, 1};
    int src_strides[ndims] = {180, 15, 5, 1};
    int dst_strides[ndims] = {1, 1, 1, 1};
    float beta = 0.0f;
    float alpha = 1.0f;

    auto nan_prop_mode = miopenNanPropagation_t::MIOPEN_NOT_PROPAGATE_NAN;
    auto indices_mode
            = miopenReduceTensorIndices_t::MIOPEN_REDUCE_TENSOR_NO_INDICES;
    auto indices_type = miopenIndicesType_t::MIOPEN_32BIT_INDICES;
    auto op_kind = miopenReduceTensorOp_t::MIOPEN_REDUCE_TENSOR_MAX;

    void *workspace = nullptr;
    size_t workspace_size = 0;

    // Create handle
    miopenHandle_t handle {};
    status_miopen = miopenCreate(&handle);
    CHECK_MIOPEN(status_miopen, "Couldn't create handle.");

    // Create Tensor descriptor
    miopenTensorDescriptor_t src_desc;
    status_miopen = miopenCreateTensorDescriptor(&src_desc);
    CHECK_MIOPEN(status_miopen, "Couldn't create src desc.");
    miopenTensorDescriptor_t dst_desc;
    status_miopen = miopenCreateTensorDescriptor(&dst_desc);
    CHECK_MIOPEN(status_miopen, "Couldn't create dst desc.");
    miopenReduceTensorDescriptor_t reduce_desc_;
    status_miopen = miopenCreateReduceTensorDescriptor(&reduce_desc_);
    CHECK_MIOPEN(status_miopen, "Couldn't create reduce desc.");

    // Set Tensor Desciptor
    status_miopen = miopenSetTensorDescriptor(
            src_desc, miopenFloat, ndims, src_dims, src_strides);
    CHECK_MIOPEN(status_miopen, "Couldn't set src desc.");

    status_miopen = miopenSetTensorDescriptor(
            dst_desc, miopenFloat, ndims, dst_dims, dst_strides);
    CHECK_MIOPEN(status_miopen, "Couldn't set dst desc.");

    status_miopen = miopenSetReduceTensorDescriptor(reduce_desc_, op_kind,
            miopenFloat, nan_prop_mode, indices_mode, indices_type);
    CHECK_MIOPEN(status_miopen, "Couldn't set reduce desc.");

    const int src_nelems = std::accumulate(
            src_dims, src_dims + ndims, 1, std::multiplies<int>());
    const int dst_nelems = std::accumulate(
            dst_dims, dst_dims + ndims, 1, std::multiplies<int>());

    float *src;
    float *dst;
    status_hip = hipMallocManaged(
            &src, src_nelems * sizeof(float), hipMemAttachGlobal);
    CHECK_HIP(status_hip, "Couldn't allocate src0.");
    status_hip = hipMallocManaged(
            &dst, dst_nelems * sizeof(float), hipMemAttachGlobal);
    CHECK_HIP(status_hip, "Couldn't allocate dst.");

    status_miopen = miopenGetReductionWorkspaceSize(
            handle, reduce_desc_, src_desc, dst_desc, &workspace_size);
    status_hip = hipMalloc((void **)&workspace, workspace_size);
    CHECK_HIP(status_hip, "Couldn't allocate workspace.");

    status_miopen = miopenReduceTensor(handle, reduce_desc_, nullptr, 0,
            workspace, workspace_size, &alpha, src_desc, src, &beta, dst_desc,
            dst);

    status_miopen = miopenDestroy(handle);
    CHECK_MIOPEN(status_miopen, "Couldn't destroy handle.");
    status_miopen = miopenDestroyTensorDescriptor(src_desc);
    CHECK_MIOPEN(status_miopen, "Couldn't destroy src desc.");
    status_miopen = miopenDestroyTensorDescriptor(dst_desc);
    CHECK_MIOPEN(status_miopen, "Couldn't destroy dst desc.");
    status_miopen = miopenDestroyReduceTensorDescriptor(reduce_desc_);
    CHECK_MIOPEN(status_miopen, "Couldn't destroy reduce desc.");
    status_hip = hipFree(src);
    CHECK_HIP(status_hip, "Couldn't free src.");
    status_hip = hipFree(dst);
    CHECK_HIP(status_hip, "Couldn't free dst.");
    status_hip = hipFree(workspace);
    CHECK_HIP(status_hip, "Couldn't free workspace.");

    return 0;
}
```

</details>

**Comment by junliume** [comment]
@qianfengz and @zjing14 
Could you take a look at this issue? Maybe an unsupported case or a potential bug in reduction implementation.

**Comment by qianfengz** [comment]
@densamoilov   
I could not reproduce the mentioned issue on ROCM.5.1 using your codes.  And you can also do verification using `[densamoilov](https://github.com/densamoilov)`

I remember the issue has been solved by some P.R.  

You can check the codes of the file [gridwise_generic_reduction_second_call_threadwise_reduce_all_dims.cpp](https://github.com/ROCmSoftwarePlatform/MIOpen/blob/develop/src/composable_kernel/composable_kernel/src/kernel_wrapper/gridwise_generic_reduction_second_call_threadwise_reduce_all_dims.cpp),   which does not have `CK_PARAM_TOREDUCE_DIMS` and `CK_PARAM_INVARIANT_DIMS` used

**Comment by densamoilov** [comment]
@qianfengz, thanks for the quick response. I have one more question just to make sure I understand how it works. Since this issue is in MIOpen and I use the latest available MIOpen release (2.16) does it mean that ROCm can include MIOpen version that is not officially released?

**Comment by qianfengz** [comment]
@densamoilov 
      `2.16` should be the major version of the MIOpen package,   you can get the complete version of MIOpen on ubuntu by `#> dpkg -s miopen-hip`.   And for each ROCM release/installation, there is a corresponding MIOpen release branch.  For example for ROCM 5.1,  the release MIOpen branch should be `release/rocm-rel-5.0`. 
      Also,  I just checked the codes [rocm-rel-5.0 gridwise_generic_reduction_second_call_threadwise_reduce_all_dims.cpp](https://github.com/ROCmSoftwarePlatform/MIOpen/blob/release/rocm-rel-5.0/src/composable_kernel/composable_kernel/src/kernel_wrapper/gridwise_generic_reduction_second_call_threadwise_reduce_all_dims.cpp),  it seems the mentioned issue have been solved by ROCM-5.0.  
     You can verify by compiling your codes against the headers provided with the package `miopen-hip`, rather than against any downloaded miopen source.  Like, 
```bash
#> hipcc -I/opt/rocm/miopen/include -L/opt/rocm/miopen/lib -lMIOpen -o test_reduce test_reduce.cpp
```

**Comment by densamoilov** [comment]
I can confirm that the issue is fixed. Thanks again for the detailed explanation!
